Feature/929 stansummary cli11

mitzimorris opened the pull request:
#### Submisison Checklist

- [x] Run tests: `./runCmdStanTests.py src/test`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Replaced use of `boost::program_options` with lib [CLI11](https://github.com/CLIUtils/CLI11),
which is a header-only C++11 library with appropriate open source license.


#### Intended Effect:

Pay down technical debt:

- Simplify build process.   Library `boost::program_options` is a non-header only library,
it made the build process longer, greatly increased the amount of the console output,

- Simplify makefiles

#### How to Verify:
Run unit tests

#### Side Effects:
NA

#### Documentation:
NA
#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

bbbales2 commented on src/test/interface/stansummary_test.cpp:
```diff
@@ -411,7 +413,7 @@ TEST(CommandStansummary, check_csv_output) {
                                 + "interface" + path_separator
                                 + "example_output" + path_separator
                                 + "tmp_test_target_csv_file.csv";
-  std::string arg_csv_file = "--csv_file=" + target_csv_file;
```
Where does the `csv_file` -> `csv_filename` change come from?

mitzimorris replied:
it was a bug in the test - the option long name is `csv_filename` - maybe the boost library was ok with prefix?

bbbales2 reviewed (commented):
The only changes here are the library, right? So it's the same features as the existing stansummary, just different code?

mitzimorris commented:
> The only changes here are the library, right? So it's the same features as the existing stansummary, just different code?

correct - the CLI11 library is a better version of boost::program_options.   updated the code to use the library; changed the unit tests to match messages generated by CLI11.

bbbales2 commented on src/cmdstan/stansummary.cpp:
```diff
@@ -36,131 +36,68 @@ Example:  stansummary model_chain_1.csv model_chain_2.csv
+      ->check(CLI::PositiveNumber);
+  app.add_option("--csv_filename,-c", csv_filename,
+                 "Write statistics to a csv.", true)
+      ->check(CLI::NonexistentPath);
```
This checks that the output file is in a writeable directory?

mitzimorris replied:
line 68 just checks that you're not clobbering an existing file. 
lines 92-100 check that the file is writeable.

looks like C++17 will have package "filesystem" that will do a nicer job on these checks.

bbbales2 commented on src/cmdstan/stansummary.cpp:
```diff
@@ -36,131 +36,68 @@ Example:  stansummary model_chain_1.csv model_chain_2.csv
+                 true);
+  app.add_option("input_files", filenames, "Sampler csv files.", true)
+      ->required()
+      ->check(CLI::ExistingFile);
```
This checks the input files are readable?

mitzimorris replied:
checks that they exist.
need to add checks for readibility, similar to check for writable output csv file.

mitzimorris replied:
done

bbbales2 replied:
Totally optional, but when I get the error for readability I don't get the friendly `--help` prompt:

```
bbales2@tadpole:~/cmdstan-develop$ bin/stansummary output4.csv
Cannot read input csv file: output4.csv.
```

Compare this to:

```
bbales2@tadpole:~/cmdstan-develop$ bin/stansummary output12321.csv
input_files: File does not exist: output12321.csv
Run with --help for more information.
```

bbbales2 replied:
Also I don't get the --help prompt with the percentiles option:

```
bbales2@tadpole:~/cmdstan-develop$ bin/stansummary -p -5 output4.csv
Option --percentiles -5: values must be in range (1,99), inclusive, and strictly increasing.
```

Very optional to add this.

mitzimorris replied:
can't figure out how to plug in custom help message via CLI11 - spent several hours looking.
will give it another go.

bbbales2 replied:
Eh if it's not easy let's not worry about it.

mitzimorris replied:
doable but not easy.  worth doing as part of next PR.

bbbales2 commented on src/test/utility.hpp:
```diff
@@ -131,7 +131,8 @@ run_command_output run_command(std::string command) {
   using boost::posix_time::ptime;
   FILE *in;
-  in = popen(command.c_str(), "r");
+  std::string command_plus = command + " 2>&1";  // put stderr to stdout
```
Why does this pipe stderr to stdout?

mitzimorris replied:
see https://stackoverflow.com/questions/6900577/c-popen-wont-catch-stderr
CLI11 lib writes arg parse errors to stderr; in order to check error message, we need stderr output as well.

bbbales2 commented on src/cmdstan/stansummary_helper.hpp:
```diff
@@ -301,11 +300,9 @@ Eigen::VectorXd percentiles_to_probs(
-      throw boost::program_options::error(message_stream.str());
+      throw std::invalid_argument(
+          "values must be in range (1,99)"
+          ", inclusive, and strictly increasing.");
```
I think there needs to be a newline after the error. Dunno where to add this to the code, but when I hit the error my cursor ends up way over at the right:

```
bbales2@tadpole:~/cmdstan-develop$ ./bin/stansummary --percentiles=-10,3,4 output2.csv
Option --percentiles -10,3,4: values must be in range (1,99), inclusive, and strictly increasing.bbales2@tadpole:~/cmdstan-develop$ 
```

mitzimorris replied:
leaving this message the same, added newline at point where error is reported - L 89 stansummary.cpp

bbbales2 commented on src/test/interface/stansummary_test.cpp:
```diff
@@ -289,14 +289,15 @@ TEST(CommandStansummary, bad_sig_figs_arg) {
 
 TEST(CommandStansummary, bad_autocorr_arg) {
-  std::string expected_message = "Bad value for option --autocorr";
+  std::string expected_message = "--autocorr: Number less or equal to 0";
```
If I don't enter an integer I get a not-super informative error:

```
bbales2@tadpole:~/cmdstan-develop$ ./bin/stansummary --autocorr=0.1 output2.csv
Could not convert: --autocorr = 0.1
Run with --help for more information.
```

Is there a way to get a better one? If not it's fine.

mitzimorris replied:
excellent question.
looks like CLI11 "master" branch has additional "TypeValidator" but not yet released.  fuck'em.
could implement a workaround - assign to string, check positive int.  PITA.

bbbales2 replied:
Skip it is fine with me. `--help` for more information is good enuff.

bbbales2 commented on src/cmdstan/stansummary_helper.hpp:
```diff
@@ -332,6 +329,12 @@ stan::mcmc::chains<> parse_csv_files(const std::vector<std::string> &filenames,
   std::ifstream ifstream;
   ifstream.open(filenames[0].c_str());
   stan::io::stan_csv stan_csv = stan::io::stan_csv_reader::parse(ifstream, out);
+  if (stan_csv.samples.rows() < 1) {
```
I deleted all the samples in a file and got a segfault:

```
bbales2@tadpole:~/cmdstan-develop$ ./bin/stansummary output2.csv
Warning: non-fatal error reading samples
terminate called after throwing an instance of 'std::invalid_argument'
  what():  No sampling draws found in Stan CSV file: output2.csv.
Aborted (core dumped)
```

bbbales2 replied:
Similarly if I deleted the csv header I get a segfault:

```
bbales2@tadpole:~/cmdstan-develop$ ./bin/stansummary output3.csv
Warning: non-fatal error reading metadata
Error: error reading header
terminate called after throwing an instance of 'std::invalid_argument'
  what():  Error with header of input file in parse
Aborted (core dumped)
```

Seems like the errors are getting detected but there's still something weird happening with memory.

mitzimorris replied:
good catch.  looking into it now.

mitzimorris replied:
OK, there was an uncaught exception.  tightened up file open/close, added more checks.

bbbales2 reviewed (changes requested):
Review!

mitzimorris commented:
hi Ben, thanks for the thorough arg checking - this code should be far more robust.

also extremely useful w/r/t this PR:  https://github.com/stan-dev/cmdstan/pull/925
there's a lot more checks that could/should be done for that PR, but it would be cool to get it in -
not sure I'll be able to get to it for this release.

mitzimorris commented:
@bbbales2 - ready for re-review

bbbales2 reviewed (approved):
Looks good to me. Merge if you're happy with it. I don't think it's worth fussing over the `--help` thing if you've already spent time trying to get that working.
